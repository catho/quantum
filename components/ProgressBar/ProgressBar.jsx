import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';
import { components } from '../shared/theme';
import { hexToRgba } from '../../shared/helpers';

const Bar = styled.div`
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;

  ${({ height, width, background }) => `
    background-color: ${hexToRgba(background, 0.39)};
    height: ${height};
    width: ${width}; 
  `}
`;

const Content = styled.div`
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  transition: width 0.4s linear;

  ${({ background, progress }) => `
    background-color: ${background};
    width: ${progress}%;
  `}
`;

const Wrapper = styled.div`
  display: flex;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  letter-spacing: 0.18px;
  line-height: 1.5;
`;

const ProgressLabel = styled.span`
  ${({ textColor }) => `
    color: ${textColor};
  `}
`;

ProgressLabel.displayName = 'ProgressLabel';

const ProgressBar = props => {
  const {
    width,
    height,
    progressText,
    progressPercent,
    label,
    skin,
    theme: {
      components: {
        progressBar: {
          skins: {
            [skin]: { background, textColor },
          },
        },
      },
    },
  } = props;

  const progressLabel = progressText === null ? progressPercent : progressText;

  return (
    <Wrapper>
      <Bar
        background={background}
        height={height}
        width={width}
        role="progressbar"
      >
        <Content background={background} progress={progressPercent} />
      </Bar>
      <ProgressLabel textColor={textColor}>
        {progressLabel}
        {label}
      </ProgressLabel>
    </Wrapper>
  );
};

ProgressBar.defaultProps = {
  width: '250px',
  height: '8px',
  progressText: null,
  progressPercent: 0,
  skin: 'primary',
  theme: {
    components: {
      progressBar: components.progressBar,
    },
  },
  label: '%',
};

ProgressBar.propTypes = {
  skin: PropTypes.oneOf(['neutral', 'primary', 'secondary']),
  theme: PropTypes.shape({
    components: PropTypes.shape({
      progressBar: PropTypes.object,
    }),
  }),
  width: PropTypes.string,
  height: PropTypes.string,
  progressText: PropTypes.number,
  progressPercent: PropTypes.number,
  label: PropTypes.string,
};

export default ProgressBar;
