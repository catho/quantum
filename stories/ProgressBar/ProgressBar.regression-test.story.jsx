import React from 'react';
import { storiesOf } from '@storybook/react';
import { ProgressBar } from '../../components';

const progressPercent = 25;
const progressText = 25;
const label = '% percent to finish';

storiesOf('ProgressBar', module)
  .add('neutral default', () => (
    <ProgressBar
      skin="neutral"
      progressPercent={progressPercent}
      progressText={progressText}
    />
  ))
  .add('neutral with label', () => (
    <ProgressBar
      skin="neutral"
      progressPercent={progressPercent}
      progressText={progressText}
      label={label}
    />
  ))
  .add('primary with label', () => (
    <ProgressBar
      skin="primary"
      progressPercent={progressPercent}
      progressText={progressText}
      label={label}
    />
  ))
  .add('secondary with label', () => (
    <ProgressBar
      skin="secondary"
      progressPercent={progressPercent}
      progressText={progressText}
      label={label}
    />
  ))
  .add('primary with differents values and percent', () => (
    <ProgressBar
      skin="primary"
      progressPercent={progressPercent}
      progressText={305}
      label=" files loaded from 1220 total"
    />
  ));
