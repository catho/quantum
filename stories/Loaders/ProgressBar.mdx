import { Meta, Story, Canvas, ArgsTable } from '@storybook/addon-docs';

import Header from '../shared/Header';
import ProgressBar from '../../components/ProgressBar';
import { Default, WithProgressPercent, WithProgressText, WithLabel} from './ProgressBar.stories.jsx';

<Meta title="Loaders/ProgressBar" component={ProgressBar} />

<Header title="ProgressBar">
    ProgressBars are used to give the user feedback of the progression of a process or an action.
</Header>

- [Importing and usage](#importing)
- [API](#api)
- [Examples](#examples)

## <a id="importing"></a>Importing and usage

```js
import ProgressBar from '@catho/quantum/ProgressBar';
```

<Canvas>
  <Story id="loaders-progressbar--default"/>
</Canvas>

## <a id="api"></a>API

<ArgsTable of={ProgressBar} />

## <a id="examples"></a>Examples

### With progress percent

<Canvas>
  <Story id="loaders-progressbar--with-progress-percent"/>
</Canvas>

### With progress text

<Canvas>
  <Story id="loaders-progressbar--with-progress-text"/>
</Canvas>


### With label

<Canvas>
  <Story id="loaders-progressbar--with-label"/>
</Canvas>
