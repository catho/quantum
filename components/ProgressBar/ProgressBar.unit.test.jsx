import React from 'react';
import renderer from 'react-test-renderer';
import { mount } from 'enzyme';
import ProgressBar from './ProgressBar';

describe('ProgressBar component', () => {
  it('Should match the snapshot of a simple progress bar', () => {
    const simpleProgressBar = <ProgressBar />;
    expect(renderer.create(simpleProgressBar).toJSON()).toMatchSnapshot();
  });
  it('Should match the snapshot of a progress bar in progress with same progressText', () => {
    const simpleProgressBar = <ProgressBar progressPercent={30} />;
    expect(renderer.create(simpleProgressBar).toJSON()).toMatchSnapshot();
  });
  it('Should match the snapshot of a progress bar in progress with same progressText', () => {
    const simpleProgressBar = (
      <ProgressBar progressPercent={30} progressText={30} />
    );
    expect(renderer.create(simpleProgressBar).toJSON()).toMatchSnapshot();
  });
  it('Should match the snapshot of a progressbar in progress with custom label', () => {
    const simpleProgressBar = (
      <ProgressBar
        progressPercent={30}
        progressText={30}
        label=" percent loaded"
      />
    );
    expect(renderer.create(simpleProgressBar).toJSON()).toMatchSnapshot();
  });
  it('Should match the snapshot of a progressbar in progress with custom label and progress text', () => {
    const simpleProgressBar = (
      <ProgressBar
        progressPercent={30}
        progressText={300}
        label=" files loaded from 1000"
      />
    );
    expect(renderer.create(simpleProgressBar).toJSON()).toMatchSnapshot();
  });

  it('should the label match with de props and shows correctly', () => {
    const label = ' percent processed';
    const component = mount(<ProgressBar progressPercent={30} label={label} />);
    const progressBarLabel = component.find('ProgressLabel').text();
    const labelShouldBe = `30${label}`;
    expect(progressBarLabel).toBe(labelShouldBe);
  });
});
