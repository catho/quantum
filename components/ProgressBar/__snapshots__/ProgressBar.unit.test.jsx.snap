// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProgressBar component Should match the snapshot of a progress bar in progress with same progressText 1`] = `
.c1 {
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;
  background-color: rgba(18,80,196,0.39);
  height: 8px;
  width: 250px;
}

.c2 {
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: width 0.4s linear;
  transition: width 0.4s linear;
  background-color: #1250C4;
  width: 30%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0.18px;
  -moz-letter-spacing: 0.18px;
  -ms-letter-spacing: 0.18px;
  letter-spacing: 0.18px;
  line-height: 1.5;
}

.c3 {
  color: #424242;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="8px"
    role="progressbar"
    width="250px"
  >
    <div
      className="c2"
    />
  </div>
  <span
    className="c3"
  >
    30
    %
  </span>
</div>
`;

exports[`ProgressBar component Should match the snapshot of a progress bar in progress with same progressText 2`] = `
.c1 {
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;
  background-color: rgba(18,80,196,0.39);
  height: 8px;
  width: 250px;
}

.c2 {
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: width 0.4s linear;
  transition: width 0.4s linear;
  background-color: #1250C4;
  width: 30%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0.18px;
  -moz-letter-spacing: 0.18px;
  -ms-letter-spacing: 0.18px;
  letter-spacing: 0.18px;
  line-height: 1.5;
}

.c3 {
  color: #424242;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="8px"
    role="progressbar"
    width="250px"
  >
    <div
      className="c2"
    />
  </div>
  <span
    className="c3"
  >
    30
    %
  </span>
</div>
`;

exports[`ProgressBar component Should match the snapshot of a progressbar in progress with custom label 1`] = `
.c1 {
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;
  background-color: rgba(18,80,196,0.39);
  height: 8px;
  width: 250px;
}

.c2 {
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: width 0.4s linear;
  transition: width 0.4s linear;
  background-color: #1250C4;
  width: 30%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0.18px;
  -moz-letter-spacing: 0.18px;
  -ms-letter-spacing: 0.18px;
  letter-spacing: 0.18px;
  line-height: 1.5;
}

.c3 {
  color: #424242;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="8px"
    role="progressbar"
    width="250px"
  >
    <div
      className="c2"
    />
  </div>
  <span
    className="c3"
  >
    30
     percent loaded
  </span>
</div>
`;

exports[`ProgressBar component Should match the snapshot of a progressbar in progress with custom label and progress text 1`] = `
.c1 {
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;
  background-color: rgba(18,80,196,0.39);
  height: 8px;
  width: 250px;
}

.c2 {
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: width 0.4s linear;
  transition: width 0.4s linear;
  background-color: #1250C4;
  width: 30%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0.18px;
  -moz-letter-spacing: 0.18px;
  -ms-letter-spacing: 0.18px;
  letter-spacing: 0.18px;
  line-height: 1.5;
}

.c3 {
  color: #424242;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="8px"
    role="progressbar"
    width="250px"
  >
    <div
      className="c2"
    />
  </div>
  <span
    className="c3"
  >
    300
     files loaded from 1000
  </span>
</div>
`;

exports[`ProgressBar component Should match the snapshot of a simple progress bar 1`] = `
.c1 {
  border-radius: 8px;
  margin-top: 8px;
  overflow: hidden;
  position: relative;
  background-color: rgba(18,80,196,0.39);
  height: 8px;
  width: 250px;
}

.c2 {
  border-radius: 8px;
  display: inline-block;
  height: 100%;
  left: 0;
  position: absolute;
  top: 0;
  -webkit-transition: width 0.4s linear;
  transition: width 0.4s linear;
  background-color: #1250C4;
  width: 0%;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  font-stretch: normal;
  font-style: normal;
  font-weight: normal;
  -webkit-letter-spacing: 0.18px;
  -moz-letter-spacing: 0.18px;
  -ms-letter-spacing: 0.18px;
  letter-spacing: 0.18px;
  line-height: 1.5;
}

.c3 {
  color: #424242;
}

<div
  className="c0"
>
  <div
    className="c1"
    height="8px"
    role="progressbar"
    width="250px"
  >
    <div
      className="c2"
    />
  </div>
  <span
    className="c3"
  >
    0
    %
  </span>
</div>
`;
